pkg/api/endpointslice/util.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpointslice
+
+import (
+	"errors"
+	"reflect"
+
+	discovery "k8s.io/api/discovery/v1alpha1"
+	"k8s.io/apimachinery/pkg/labels"
+	discoverylisters "k8s.io/client-go/listers/discovery/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+)
+
+// GetMatchingServiceSubset returns Endpoints for any EndpointSlices with
+// matching Ports, AddressType, and Service. This data should closely parallel
+// data from a corresponding v1.EndpointSubset.
+func GetMatchingServiceSubset(esl discoverylisters.EndpointSliceLister, endpointSliceKey string) ([]discovery.Endpoint, error) {
+	labelKeys := []string{discovery.LabelServiceName}
+	return GetMatchingSubset(esl, endpointSliceKey, labelKeys)
+}
+
+// GetMatchingSubset returns Endpoints for any EndpointSlices with matching
+// Ports, AddressType, and specified labels. This data should closely parallel
+// data from a corresponding v1.EndpointSubset.
+func GetMatchingSubset(esl discoverylisters.EndpointSliceLister, endpointSliceKey string, labelKeys []string) ([]discovery.Endpoint, error) {
+	if len(labelKeys) == 0 {
+		return nil, errors.New("Must provide labelKeys to filter EndpointSlices with")
+	}
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(endpointSliceKey)
+	if err != nil {
+		return nil, err
+	}
+
+	endpointSlice, err := esl.EndpointSlices(namespace).Get(name)
+	if err != nil {
+		return nil, err
+	}
+
+	labelSet := labels.Set{}
+	for _, key := range labelKeys {
+		labelSet[key] = endpointSlice.Labels[key]
+	}
+	selector := labels.SelectorFromSet(labels.Set(labelSet))
```

Comment:
this construction would be nice to avoid if possible. can we pass in a selector instead, and make the default selector used by GetMatchingServiceSubset a package var we allocate once?